File: intellij-plugin/src/main/kotlin/dev/skillbill/intellij/infrastructure/cli/CliSkillBillStatusRepository.kt
package dev.skillbill.intellij.infrastructure.cli

import com.intellij.openapi.application.ApplicationManager
import dev.skillbill.intellij.application.PreferenceCachePort
import dev.skillbill.intellij.application.StatusRepository
import dev.skillbill.intellij.domain.DEFAULT_CLI_TIMEOUT_MS
import dev.skillbill.intellij.domain.DEFAULT_STDERR_LIMIT_BYTES
import dev.skillbill.intellij.domain.DEFAULT_STDOUT_LIMIT_BYTES
import dev.skillbill.intellij.domain.SkillBillStatusOutcome
import dev.skillbill.intellij.domain.StatusClock
import dev.skillbill.intellij.domain.StatusDiagnostic
import dev.skillbill.intellij.domain.UnavailableReason
import java.nio.file.Path

/**
 * CLI-backed [StatusRepository]. Runs
 * `skill-bill work status --repo-root <canonical> --format json` off the EDT.
 */
class CliSkillBillStatusRepository(
    private val preferences: PreferenceCachePort,
    private val processRunner: ProcessRunner,
    private val clock: StatusClock = StatusClock.system(),
    private val executableResolver: () -> CliExecutableResolution = { CliExecutableResolver.resolve(preferences) },
    private val timeoutMs: Long = DEFAULT_CLI_TIMEOUT_MS,
    private val stdoutLimitBytes: Int = DEFAULT_STDOUT_LIMIT_BYTES,
    private val stderrLimitBytes: Int = DEFAULT_STDERR_LIMIT_BYTES,
) : StatusRepository {
    override suspend fun fetchStatus(projectRoot: Path): SkillBillStatusOutcome {
        val observedAt = clock.now()
        val executable = when (val resolution = executableResolver()) {
            is CliExecutableResolution.Found -> resolution.path
            CliExecutableResolution.Missing ->
                return SkillBillStatusOutcome.Unavailable(
                    observedAt = observedAt,
                    summary = MISSING_EXECUTABLE_SUMMARY,
                    reasonCode = UnavailableReason.MISSING_EXECUTABLE,
                    diagnostic = StatusDiagnostic(reasonCode = "missing_executable"),
                )
            CliExecutableResolution.Misconfigured ->
                return SkillBillStatusOutcome.Unavailable(
                    observedAt = observedAt,
                    summary = MISCONFIGURED_EXECUTABLE_SUMMARY,
                    reasonCode = UnavailableReason.MISCONFIGURED,
                    diagnostic = StatusDiagnostic(reasonCode = "misconfigured_executable"),
                )
        }
        val canonicalRoot = try {
            projectRoot.toAbsolutePath().normalize().toRealPath()
        } catch (_: Exception) {
            return SkillBillStatusOutcome.Unavailable(
                observedAt = observedAt,
                summary = "Project root is not a usable path",
                reasonCode = UnavailableReason.INVALID_REPOSITORY_INPUT,
                diagnostic = StatusDiagnostic(reasonCode = "invalid_repository_input"),
            )
        }

        val result = try {
            processRunner.runCoalesced(
                ProcessSpec(
                    command = listOf(
                        executable,
                        "work",
                        "status",
                        "--repo-root",
                        canonicalRoot.toString(),
                        "--format",
                        "json",
                    ),
                    timeoutMs = timeoutMs,
                    stdoutLimitBytes = stdoutLimitBytes,
                    stderrLimitBytes = stderrLimitBytes,
                ),
            )
        } catch (_: Exception) {
            // Never leak command output or paths from the failure surface.
            return SkillBillStatusOutcome.Unavailable(
                observedAt = observedAt,
                summary = "Skill Bill status command failed to start",
                reasonCode = UnavailableReason.PROCESS_FAILURE,
                diagnostic = StatusDiagnostic(reasonCode = "process_start_failure"),
            )
        }

        if (result.cancelled) {
            return SkillBillStatusOutcome.Unavailable(
                observedAt = observedAt,
                summary = "Skill Bill status poll cancelled",
                reasonCode = UnavailableReason.CANCELLED,
                diagnostic = StatusDiagnostic(cancelled = true, reasonCode = "cancelled"),
            )
        }
        if (result.timedOut) {
            return SkillBillStatusOutcome.Unavailable(
                observedAt = observedAt,
                summary = "Skill Bill status poll timed out",
                reasonCode = UnavailableReason.TIMEOUT,
                diagnostic = StatusDiagnostic(timedOut = true, reasonCode = "timeout"),
            )
        }
        return IdeStatusJsonMapper.map(
            stdout = result.stdout,
            observedAt = observedAt,
            exitCode = result.exitCode,
        )
    }

    companion object {
        const val MISSING_EXECUTABLE_SUMMARY: String =
            "Skill Bill CLI not found — set its path in Settings | Tools | Skill Bill"
        const val MISCONFIGURED_EXECUTABLE_SUMMARY: String =
            "Skill Bill CLI path override is not usable — check Settings | Tools | Skill Bill"

        fun intellijEdtGuard(): EdtGuard = EdtGuard {
            val app = ApplicationManager.getApplication()
            if (app != null && app.isDispatchThread) {
                error("Skill Bill status CLI must not run on the EDT")
            }
        }
    }
}

File: vscode-extension/src/infrastructure/cli/CliSkillBillStatusRepository.ts
import * as fs from "fs";
import * as path from "path";
import { PreferenceCachePort } from "../../application/PreferenceCachePort";
import { StatusRepository } from "../../application/StatusRepository";
import {
  DEFAULT_CLI_TIMEOUT_MS,
  DEFAULT_STDERR_LIMIT_BYTES,
  DEFAULT_STDOUT_LIMIT_BYTES,
} from "../../domain/Constants";
import { SkillBillStatusOutcome, UnavailableReason } from "../../domain/SkillBillStatusOutcome";
import { StatusClock } from "../../domain/StatusClock";
import { CliExecutableResolution, resolveCliExecutable } from "./CliExecutableResolver";
import { mapIdeStatusJson } from "./IdeStatusJsonMapper";
import { BoundedProcessResult, ProcessRunner, ProcessSpec } from "./ProcessRunner";

export const MISSING_EXECUTABLE_SUMMARY =
  "Skill Bill CLI not found — set skillBill.cliPath in settings";

export const MISCONFIGURED_EXECUTABLE_SUMMARY =
  "Skill Bill CLI path override is not usable — check skillBill.cliPath in settings";

export class CliSkillBillStatusRepository implements StatusRepository {
  constructor(
    private readonly preferences: PreferenceCachePort,
    private readonly processRunner: ProcessRunner,
    private readonly clock: StatusClock = StatusClock.system(),
    private readonly executableResolver: () => CliExecutableResolution = () =>
      resolveCliExecutable(this.preferences),
    private readonly timeoutMs: number = DEFAULT_CLI_TIMEOUT_MS,
    private readonly stdoutLimitBytes: number = DEFAULT_STDOUT_LIMIT_BYTES,
    private readonly stderrLimitBytes: number = DEFAULT_STDERR_LIMIT_BYTES,
  ) {}

  async fetchStatus(projectRoot: string): Promise<SkillBillStatusOutcome> {
    const observedAt = this.clock.now();
    const resolution = this.executableResolver();
    if (resolution.kind === "missing") {
      return {
        kind: "unavailable",
        observedAt,
        summary: MISSING_EXECUTABLE_SUMMARY,
        reasonCode: UnavailableReason.MISSING_EXECUTABLE,
        diagnostic: { reasonCode: "missing_executable" },
      };
    }
    if (resolution.kind === "misconfigured") {
      return {
        kind: "unavailable",
        observedAt,
        summary: MISCONFIGURED_EXECUTABLE_SUMMARY,
        reasonCode: UnavailableReason.MISCONFIGURED,
        diagnostic: { reasonCode: "misconfigured_executable" },
      };
    }

    let canonicalRoot: string;
    try {
      canonicalRoot = fs.realpathSync(path.resolve(projectRoot));
    } catch {
      return {
        kind: "unavailable",
        observedAt,
        summary: "Project root is not a usable path",
        reasonCode: UnavailableReason.INVALID_REPOSITORY_INPUT,
        diagnostic: { reasonCode: "invalid_repository_input" },
      };
    }

    let result: BoundedProcessResult;
    try {
      const spec: ProcessSpec = {
        command: [
          resolution.path,
          "work",
          "status",
          "--repo-root",
          canonicalRoot,
          "--format",
          "json",
        ],
        timeoutMs: this.timeoutMs,
        stdoutLimitBytes: this.stdoutLimitBytes,
        stderrLimitBytes: this.stderrLimitBytes,
      };
      result = await this.processRunner.runCoalesced(spec);
    } catch {
      return {
        kind: "unavailable",
        observedAt,
        summary: "Skill Bill status command failed to start",
        reasonCode: UnavailableReason.PROCESS_FAILURE,
        diagnostic: { reasonCode: "process_start_failure" },
      };
    }

    if (result.cancelled) {
      return {
        kind: "unavailable",
        observedAt,
        summary: "Skill Bill status poll cancelled",
        reasonCode: UnavailableReason.CANCELLED,
        diagnostic: { cancelled: true, reasonCode: "cancelled" },
      };
    }
    if (result.timedOut) {
      return {
        kind: "unavailable",
        observedAt,
        summary: "Skill Bill status poll timed out",
        reasonCode: UnavailableReason.TIMEOUT,
        diagnostic: { timedOut: true, reasonCode: "timeout" },
      };
    }
    return mapIdeStatusJson(result.stdout, observedAt, result.exitCode);
  }
}
